tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -592,8 +606,18 @@ SpirvFunctionParameter *
 DeclResultIdMapper::createFnParam(const ParmVarDecl *param) {
   const auto type = getTypeOrFnRetType(param);
   const auto loc = param->getLocation();
-  SpirvFunctionParameter *fnParamInstr = spvBuilder.addFnParam(
-      type, param->hasAttr<HLSLPreciseAttr>(), loc, param->getName());
+  SpirvFunctionParameter *fnParamInstr = nullptr;
+  if (isOpaqueArrayType(param->getType()) &&
+      !param->getType()->isConstantArrayType()) {
+    // If it is a bindless array of an opaque type, we have to use
+    // a pointer to a pointer of the runtime array.
+    fnParamInstr = spvBuilder.addFnParam(
+        spvContext.getPointerType(type, spv::StorageClass::UniformConstant),
+        param->hasAttr<HLSLPreciseAttr>(), loc, param->getName());
+  } else {
+    fnParamInstr = spvBuilder.addFnParam(
+        type, param->hasAttr<HLSLPreciseAttr>(), loc, param->getName());
+  }
```

Comment:
I think the above logic should be performed inside `addFnParam`. That way, anyone calling `addFunctionParam` will get the right behavior for free. Meaning:

```cpp
SpirvFunctionParameter *SpirvBuilder::addFnParam(QualType ptrType,
                                                 bool isPrecise,
                                                 SourceLocation loc,
                                                 llvm::StringRef name) {
  assert(function && "found detached parameter");
  SpirvFunctionParameter *param = nullptr;
  if(isBindlessOpaqueArray(type)) {
  // If it is a bindless array of an opaque type, we have to use
  // a pointer to a pointer of the runtime array.
  param = new (context) SpirvFunctionParameter(spvContext.getPointerType(type, spv::StorageClass::UniformConstant), isPrecise, loc);
  } else {
    param = new (context) SpirvFunctionParameter(ptrType, isPrecise, loc);
  }

  param->setStorageClass(spv::StorageClass::Function);
  param->setDebugName(name);
  function->addParameter(param);
  return param;
}
```

You also won't need to create two versions of `addFnParam` method. What do you think?

Reply:
+1. Nice catch. I would be much better. Updated.